File: 6.controllers-services-repositories-middlewares-and-request-context/code/typescript/07the_repository_layer_4.ts
export class BookRepo {
    constructor(private db: any) {}

    // ONE method, ONE result shape: all books, sorted.
    async findAllBooks(sort: string) {
        // Build the query from data passed down by the service.
        const query = `SELECT id, title, author FROM books ORDER BY ${sort}`;
        const [rows] = await this.db.execute(query);
        return rows;
    }

    // SEPARATE method for one book. No optional toggle parameter.
    async findBookById(bookId: number) {
        const query = "SELECT id, title, author FROM books WHERE id = ?";
        const [rows] = await this.db.execute(query, [bookId]);
        return rows[0];
    }
}

File: 6.controllers-services-repositories-middlewares-and-request-context/code/rust/07the_repository_layer_4.rs
use sqlx::PgPool;

pub struct BookRepo {
    db: PgPool,
}

#[derive(sqlx::FromRow)]
pub struct Book {
    pub id: i32,
    pub title: String,
    pub author: String,
}

impl BookRepo {
    // ONE method, ONE shape of result: all books, sorted.
    pub async fn find_all_books(&self, sort: &str) -> Result<Vec<Book>, sqlx::Error> {
        // Build the query from the data the service handed down.
        let query = format!("SELECT id, title, author FROM books ORDER BY {}", sort);
        let books = sqlx::query_as::<_, Book>(&query)
            .fetch_all(&self.db)
            .await?;
        Ok(books)
    }

    // A SEPARATE method for the single-book case. No optional toggles.
    pub async fn find_book_by_id(&self, id: i32) -> Result<Book, sqlx::Error> {
        let book = sqlx::query_as::<_, Book>("SELECT id, title, author FROM books WHERE id = $1")
            .bind(id)
            .fetch_one(&self.db)
            .await?;
        Ok(book)
    }
}
